### PR TITLE
Feature/album api

### PR DESCRIPTION
**Pull request by manoger**
__( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩__

# New things:
* Mugshot module
* API Tests.
* GET/POST/DELETE http methods for profile image.
* ImageUtils for random image generation.
* ImageUtils for resizing images to square shape.
# Changes:
* Max upload file size as property.
* Image dimensions as property.
* Default directory name as property.
# Removed:
* Entire module with "Album" as name.

__( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩( ᐛ )و ٩__

### CODE REVIEW

File: mugshot/src/main/java/hive/album/controller/UtilsController.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+@RequestMapping("/utils")
+public class UtilsController {
+  private final UserRepository userRepository;
+  private final ImageStorer imageStorer;
+  @Value("${hive.album.profile-image-name}")
+  private String imageName;
+
+
+  @ResponseStatus(code = HttpStatus.OK, reason = "The server is working")
+  @GetMapping("/hey") public void hey(){}
```
Comment (eiling) [inline mugshot/src/main/java/hive/album/controller/UtilsController.java:27]: Nani?

File: mugshot/src/main/java/hive/album/controller/UtilsController.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  @Autowired
+  public UtilsController(ImageStorer imageStorer,UserRepository userRepository){
+    this.userRepository=userRepository;
+    this.imageStorer=imageStorer;
+    userRepository.save(new User("Test_User","123",0,0));
```
Comment (eiling) [inline mugshot/src/main/java/hive/album/controller/UtilsController.java:32]: Remove test things from main code

File: mugshot/src/main/java/hive/album/controller/UtilsController.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  @GetMapping(value = "searchImage/{profileImageName:.+}", produces = MediaType.IMAGE_JPEG_VALUE)
+  public ResponseEntity<Resource> searchSomeImage(@RequestHeader(name = HiveHeaders.AUTHENTICATED_USER_NAME_HEADER) final String username, @PathVariable String profileImageName) {
```
Comment (eiling) [inline mugshot/src/main/java/hive/album/controller/UtilsController.java:36]: Why?

File: mugshot/src/main/java/hive/mugshot/controller/MugshotController.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
+import org.springframework.core.io.Resource;
+
+import static hive.mugshot.storage.ImageUtils.validateIfHasAnImageAsExtension;
+
+@RestController
+@RequestMapping("/mugshot")
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/controller/MugshotController.java]: This should be "/"

File: mugshot/src/main/java/hive/mugshot/controller/MugshotController.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      @RequestHeader(name = HiveHeaders.AUTHENTICATED_USER_NAME_HEADER) final String username) {
+    if(!validateIfHasAnImageAsExtension(insertedImage.getOriginalFilename())) {
+      throw new NotAcceptedFileFormatException();
+    }
+    var userID=userRepository.findByUsername(username).getId().toString();
+    imageStorer.storeImageProfile(userID,insertedImage,imageName);
+  }
+
+  @GetMapping(produces = MediaType.IMAGE_JPEG_VALUE)
+  public ResponseEntity<Resource> searchProfileImage(
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/controller/MugshotController.java:47]: Should be
```java
public ResponseEntity<Resource> searchProfileImage(
    @RequestHeader(name = HiveHeaders.AUTHENTICATED_USER_NAME_HEADER) final String username
) {
  // ...
}
```

File: mugshot/src/main/java/hive/mugshot/controller/MugshotController.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      @RequestHeader(name = HiveHeaders.AUTHENTICATED_USER_NAME_HEADER) final String username) {
+    var userID=userRepository.findByUsername(username).getId().toString();
+    Resource file = imageStorer.loadImage(userID,imageName);
+    return ResponseEntity.ok().header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + file.getFilename() + "\"").body(file);
+  }
+
+  @ResponseStatus(code = HttpStatus.NO_CONTENT, reason = "Profile image successfully deleted")
+  @DeleteMapping
+  public void deleteProfileImage(
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/controller/MugshotController.java:57]: Same as above

File: mugshot/src/main/java/hive/mugshot/controller/UtilsController.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
+package hive.mugshot.controller;
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/controller/UtilsController.java:1]: Two of the same file?

File: mugshot/src/main/java/hive/mugshot/exception/UserNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package hive.mugshot.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.BAD_REQUEST, reason = "The user does not exist")
+public class UserNotFoundException extends RuntimeException {
+}
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/exception/UserNotFoundException.java]: Insert EOL at EOF

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+
+  private void createDirectoryIfNotExist(String userDirectoryPath){
+    Path parentDir = Paths.get(rootDir,userDirectoryPath);
+    if (!Files.exists(parentDir)) {
+      try {
+        Files.createDirectories(parentDir);
+      } catch (IOException e) {
+        throw new RuntimeException("Not capable to create the directory.\n"+e);
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: not capable -> unable

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import java.nio.file.*;
+
+@Service
+public class ImageStorer {
+  @Value("${hive.mugshot.image-directory-path}")
+  private String rootDir;
+  @Value("${hive.mugshot.profile-image-dimension}")
+  private int imageSizeInPixels;
+  public void storeImageProfile(String userDirectoryName, MultipartFile insertedImage, String imageStoredName){
+    this.createDirectoryIfNotExist(userDirectoryName);
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: 'this' not needed

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    try {
+      var buff = ImageUtils.resizeImageToSquare(ImageIO.read(insertedImage.getInputStream()),imageSizeInPixels);
+      ImageIO.write(buff, "jpg", resolveRootPathFullUrlWith(userDirectoryName, imageStoredName).toFile());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+  public void storeImageProfile(String userDirectoryName, BufferedImage insertedImage, String imageStoredName){
+    this.createDirectoryIfNotExist(userDirectoryName);
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: 'this' not needed

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    try {
+      var buff = ImageUtils.resizeImageToSquare(insertedImage,imageSizeInPixels);
+      ImageIO.write(buff, "jpg", resolveRootPathFullUrlWith(userDirectoryName, imageStoredName).toFile());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+  public Resource loadImage(String userDirectoryName,String ImageName) {
+    try {
+      Path file = resolveRootPathFullUrlWith(userDirectoryName,ImageName);
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: use var

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+      Resource resource = new UrlResource(file.toUri());
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: use var

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+      }else{
+        throw new ImageNotFound();
+      }
+    } catch (MalformedURLException e) {
+      e.printStackTrace();
+      throw new InvalidPathException();
+    }
+  }
+  public void deleteImage(String userDirectoryName, String ImageName) {
+    Path parentDir = resolveRootPathFullUrlWith(userDirectoryName,ImageName);
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: use var

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    try {
+      Files.deleteIfExists(parentDir);
+    } catch (IOException e) {
+      e.printStackTrace();
+      throw new RuntimeException("Not capable to delete the directory.\n"+e);
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: not capable -> unable

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    Path parentDir = Paths.get(rootDir,userDirectoryPath);
+    if (!Files.exists(parentDir)) {
+      try {
+        Files.createDirectories(parentDir);
+      } catch (IOException e) {
+        throw new RuntimeException("Not capable to create the directory.\n"+e);
+      }
+    }
+  }
+  private Path resolveRootPathFullUrlWith(String userDirectoryName, String filename) {
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: What does this do?

File: mugshot/src/main/java/hive/mugshot/storage/ImageUtils.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
+package hive.mugshot.storage;
+
+import java.awt.*;
+import java.awt.image.BufferedImage;
+import java.util.Random;
+import java.util.regex.Pattern;
+
+public final class ImageUtils {
+  private static final String IMAGE_PATTERN = "(.+\\.(gif|png|bmp|jpeg|jpg)$)";
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageUtils.java]: Insert start of line in regex
```(^.+\\.(gif|png|bmp|jpeg|jpg)$)```

File: mugshot/src/main/resources/application.properties [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+#spring.jpa.hibernate.ddl-auto=create-drop
+spring.jpa.hibernate.ddl-auto=create
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+logging.level.org.hibernate.type=trace
+# Mugshot
+hive.mugshot.image-directory-path=./mugshot_images_profiles
+hive.mugshot.profile-image-name=ProfileImage.jpg
+hive.mugshot.profile-image-dimension=256
+spring.http.multipart.max-file-size=1MB
```
Comment (eiling) [inline mugshot/src/main/resources/application.properties]: Insert EOL at EOF

File: mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  @Test
+  public void givenValidImage_WhenImageDeleted_then204isReturned() throws Exception{
+    mockMvc.perform(
+        delete("/mugshot").header("authenticated-user-name",username)
+    ).andExpect(status().isNoContent());
+  }
+
+  //ERRORS:
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java]: ?

File: mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+    MockitoAnnotations.initMocks(this);
+    when(user.getId()).thenReturn(userId);
+    when(userRepository.findByUsername(username)).thenReturn(user);
+    var mugshotController = new MugshotController(imageStorer,userRepository);
+    ReflectionTestUtils.setField(mugshotController, "imageName", imageName);
+    mockMvc = MockMvcBuilders.standaloneSetup(mugshotController).build();
+    validDirectoryName = (rootDir + "/" + user.getId() + "/");
+  }
+
+  //SUCCESS:
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java]: ?

File: mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+    createDirectoryForTest(validDirectoryName);
+    var resourceImage=createImageForTest(validDirectoryName);
+    given(imageStorer.loadImage(userId.toString(),imageName))
+        .willReturn(resourceImage);
+      mockMvc.perform(
+          get("/mugshot")
+          .header("authenticated-user-name", username))
+          .andExpect(status().isOk())
+          .andExpect(content().contentType(MediaType.IMAGE_JPEG))
+          .andDo(print());
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java:88]: No need to print the test results.

File: mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+      multipartFile = new MockMultipartFile("image", "Unsupported.Extension.wmv", MediaType.APPLICATION_PDF_VALUE, byteArrayOutputStream.toByteArray());
+      mockMvc.perform(multipart("/mugshot")
+          .file(multipartFile)
+          .header("authenticated-user-name", username))
+          .andExpect(status().isUnsupportedMediaType())
+          .andDo(print());
+  }
+
+  @Test
+  public void givenWrongBodyRequestKey_WhenImageUploaded_then400isReturned() throws Exception{
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/controller/MugshotControllerTest.java]: request body, not body request

File: mugshot/src/test/java/hive/mugshot/controller/UtilsControllerTest.java [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+    ReflectionTestUtils.setField(utilsController, "imageName", imageName);
+    mockMvc = MockMvcBuilders.standaloneSetup(utilsController).build();
+  }
+
+  @Test
+  public void givenRequestWithOkResponseAsDefault_whenVerifyIfTheServeIsOk_then200isReturned() throws Exception{
+    mockMvc
+        .perform(
+            get("/utils/hey")
+        )
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/controller/UtilsControllerTest.java]: ).andExpect(status().isOk());

File: mugshot/src/test/java/hive/mugshot/storage/ImageStorerTest.java [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+    multipartFile=new MockMultipartFile(
+        "image",
+        "WhateverName.gif",
+        MediaType.IMAGE_JPEG_VALUE,
+        byteArrayOutputStream.toByteArray()
+    );
+  }
+
+  @Test
+  public void uploadImage_WhenImageNameAndUserDirectoryIsNotNull_expectExistenceOfFile(){
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/storage/ImageStorerTest.java]: expectFileExists
no need to be poetic

File: mugshot/src/test/java/hive/mugshot/storage/ImageStorerTest.java [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+    imageStorer.storeImageProfile(userId.toString(),multipartFile,imageName);
+    Path path=Paths.get(rootDir,userId.toString(),imageName);
+    Assert.assertTrue(Files.exists(path));
+  }
+
+  @Test
+  public void deleteImage_WhenImageNameAndUserDirectoryIsNotNull_expectNoExistenceOfFile(){
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/storage/ImageStorerTest.java]: expectFileDoesNotExist

File: mugshot/src/test/java/hive/mugshot/storage/ImageStorerTest.java [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+        MediaType.IMAGE_JPEG_VALUE,
+        byteArrayOutputStream.toByteArray()
+    );
+  }
+
+  @Test
+  public void uploadImage_WhenImageNameAndUserDirectoryIsNotNull_expectExistenceOfFile(){
+    imageStorer.storeImageProfile(userId.toString(),multipartFile,imageName);
+    Path path=Paths.get(rootDir,userId.toString(),imageName);
+    Assert.assertTrue(Files.exists(path));
```
Comment (eiling) [inline mugshot/src/test/java/hive/mugshot/storage/ImageStorerTest.java]: import static Assert

File: mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+  public void storeImageProfile(String userDirectoryName, BufferedImage insertedImage, String imageStoredName){
+    this.createDirectoryIfNotExist(userDirectoryName);
+    try {
+      var buff = ImageUtils.resizeImageToSquare(insertedImage,imageSizeInPixels);
+      ImageIO.write(buff, "jpg", resolveRootPathFullUrlWith(userDirectoryName, imageStoredName).toFile());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+  public Resource loadImage(String userDirectoryName,String ImageName) {
```
Comment (eiling) [inline mugshot/src/main/java/hive/mugshot/storage/ImageStorer.java]: Every image has the same name so...
Also cammelCase

File: settings.gradle [hunk 28]
```diff
@@ -5,4 +5,5 @@ include 'common'
 include 'gateway'
 include 'naming-server'
 include 'entity'
+include 'album'
```
Comment (eiling) [inline settings.gradle]: mugshot?

**Review by eiling — CHANGES_REQUESTED** [review]
- Fix spacing
- OK responses don't need a reason?
- Use var

**Comment by manoger** [comment]
I'll refactor.

**Pull request closed without merging**